ms2rescore/feature_generators/ms2.py

```
@@ -0,0 +1,319 @@
+        """Retrieve annotated spectra for all psms."""
+
+        spectrum_reader = self._read_spectrum_file(spectrum_file)
+
+        spectrum_id_pattern = re.compile(
+            self.spectrum_id_pattern if self.spectrum_id_pattern else r"(.*)"
+        )
+
+        try:
+            mapper = {
+                spectrum_id_pattern.search(spectrum_id).group(1): spectrum_id
+                for spectrum_id in spectrum_reader._offset_index.mapping["spectrum"].keys()
+            }
+        except AttributeError:
+            raise ParseSpectrumError(
+                "Could not parse spectrum IDs using ´spectrum_id_pattern´. Please make sure that there is a capturing in the pattern."
+            )
+        annotated_spectra = [
+            self._annotate_spectrum(psm, spectrum_reader.get_by_id(mapper[psm.spectrum_id]))
+            for psm in psm_list
+        ]
+        for psm, annotated_spectrum in zip(psm_list, annotated_spectra):
+            psm.rescoring_features.update(
+                self._calculate_spectrum_features(psm, annotated_spectrum)
+            )
+
+            if self.calculate_hyperscore:
+                # Filters out peaks which are unnannotated (can be specified, but keep at b-y ions of any charge ?)
+                b, y = get_by_fragments(annotated_spectrum)
+                hs = calculate_hyperscore(
+                    n_y=len(y),
+                    n_b=len(b),
+                    y_ion_intensities=y,
+                    b_ion_intensities=b
+                )
+                psm.rescoring_features.update({'hyperscore': hs})
+
+    @staticmethod
+    def _read_spectrum_file(spectrum_filepath: Path) -> Union[mzml.PreIndexedMzML, mgf.IndexedMGF]:
+
+        if spectrum_filepath.suffix.lower() == ".mzml":
+            return mzml.PreIndexedMzML(str(spectrum_filepath))
+        elif spectrum_filepath.suffix.lower() == ".mgf":
+            return mgf.IndexedMGF(str(spectrum_filepath))
```

It might be better to avoid failing silently and add an else and raise an e.g. `NotImplementedError` or `ValueError`.